File: src/main/java/eu/playperium/spongecore/configuration/ConfigurationManager.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package eu.playperium.spongecore.configuration;

import ninja.leaping.configurate.commented.CommentedConfigurationNode;
import ninja.leaping.configurate.loader.ConfigurationLoader;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;

public class ConfigurationManager {

    private File configurationFile;

    private ConfigurationLoader<CommentedConfigurationNode> configurationLoader;
    private CommentedConfigurationNode configurationNode = null;

    private ArrayList<ConfigOption> configOptions = new ArrayList<>();

    public ConfigurationManager(File configurationFile, ConfigurationLoader<CommentedConfigurationNode> configurationLoader) {
        this.configurationFile = configurationFile;
        this.configurationLoader = configurationLoader;

        try {
            this.configurationNode = configurationLoader.load();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void registerConfigOption(ConfigOption option) {
        configOptions.add(option);
    }

    public void setConfigOptions() {
        if (!configurationFile.exists()) {
            try {
                configurationFile.createNewFile();

                for (ConfigOption option: configOptions) {
                    if (option.getDefaultValue() != null) {
                        configurationNode.getNode(option.getPath()).setValue(option.getDefaultValue());
                        configurationNode.getNode();
                    }
                }

                configurationLoader.save(this.configurationNode);

            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public String getString(ConfigOption option) {
        return configurationNode.getNode(option.getPath()).getString();
    }

    public float getFloat(ConfigOption option) {
        return configurationNode.getNode(option.getPath()).getFloat();
    }

    public double getDouble(ConfigOption option) {
        return configurationNode.getNode(option.getPath()).getDouble();
    }

    public int getInt(ConfigOption option) {
        return configurationNode.getNode(option.getPath()).getInt();
    }

    public long getLong(ConfigOption option) {
        return configurationNode.getNode(option.getPath()).getLong();
    }

    public boolean getBoolean(ConfigOption option) {
        return configurationNode.getNode(option.getPath()).getBoolean();
    }

    public void setValue(ConfigOption option, Object value) {
        configurationNode.getNode(option.getPath()).setValue(value);

        try {
            configurationLoader.save(this.configurationNode);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/eu/playperium/spongecore/configuration/ConfigurationManager.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package eu.playperium.spongecore.configuration

import ninja.leaping.configurate.commented.CommentedConfigurationNode
import ninja.leaping.configurate.loader.ConfigurationLoader
import java.io.File
import java.util.*

class ConfigurationManager(configurationFile: File, configurationLoader: ConfigurationLoader<CommentedConfigurationNode>) {

    private val configurationFile: File = configurationFile
    private val configurationLoader: ConfigurationLoader<CommentedConfigurationNode> = configurationLoader
    private val configurationNode: CommentedConfigurationNode = configurationLoader.load()
    private val configOptions: ArrayList<ConfigOption> = ArrayList()

    fun registerConfigOption(option: ConfigOption) {
        if (!configOptions.contains(option)) {
            configOptions.add(option)
        }
    }

    fun setConfigOptions() {
        if (!configurationFile.exists()) {
            configurationFile.createNewFile()

            for (option: ConfigOption in configOptions) {
                configurationNode.getNode(*option.path).value = option.defaultValue
            }

            configurationLoader.save(this.configurationNode)
        }
    }

    fun setValue(option: ConfigOption, value: Any) {
        this.configurationNode.getNode(*option.path).value = value
        configurationLoader.save(this.configurationNode)
    }

    fun getString (option: ConfigOption): String {
        return this.configurationNode.getNode(*option.path).string
    }

    fun getFloat(option: ConfigOption): Float {
        return this.configurationNode.getNode(*option.path).float
    }

    fun getDouble(option: ConfigOption): Double {
        return this.configurationNode.getNode(*option.path).double
    }

    fun getInt(option: ConfigOption): Int {
        return this.configurationNode.getNode(*option.path).int
    }

    fun getLong(option: ConfigOption): Long {
        return this.configurationNode.getNode(*option.path).long
    }

    fun getBoolean(option: ConfigOption): Boolean {
        return this.configurationNode.getNode(*option.path).boolean
    }
}
